### ISSUE TITLE
Use extracted google-discovery plugin

### ISSUE THREAD

**Issue by hamnis** [armanbilge/gcp4s#254]
I have extracted the Discovery plugin [here](https://github.com/hamnis/google-discovery-scala) and made it work with 2.12, 2.13 and 3. (publishing of this plugin is still pending)

I have reimplemented the codegeneration using [Paiges](https://github.com/typelevel/paiges), but the initial code by @armanbilge  is still there.

I have also added a basic client generation as well. However with no google authentication, as I think that should be an http4s middleware, which I think this library has.

**Comment by armanbilge** [comment]
Hi, thanks for sharing!

---

> I have also added a basic client generation as well.

Recently I've been experimenting with using Guardrail to generate clients.

There is a project that converts Google Discovery APIs to OpenAPI specs:
https://github.com/APIs-guru/openapi-directory/tree/main/APIs/googleapis.com

These can then be fed into Guardrail.
https://github.com/guardrail-dev/guardrail

---

> However with no google authentication, as I think that should be an http4s middleware, which I think this library has.

Yes! Actually I've been working on extracting this into a new runtime library, that supports JVM/JS/Native.

https://github.com/davenverse/googleapis-http4s-runtime

Besides these Discovery APIs, the goal is to provided runtime support for the generated gRPC APIs published here:

https://github.com/davenverse/googleapis-http4s

**Comment by hamnis** [comment]
I am not a fan of the guardrail generated code. So I would still do this. 
But nice that someone has done the work to convert it into openapi specs. 
I was thinking on going in that direction.

our end goal is to help build [this](https://github.com/nrkno/bigquery-scala/) in pure scala.
And that requires us to also look into the googleapis-http4s grpc stuff.

Maybe our goals are somewhat aligned here.

**Comment by armanbilge** [comment]
> I am not a fan of the guardrail generated code. So I would still do this.

Perhaps I'm missing something, but aren't you unnecessary replicating that toolchain?

What I mean is, even if you don't like the Guardrail-generated code (I'm not crazy about it either), a better solution to that might be to build a new generator for Guardrail that generates code that you do like? At least it can reuse the parsing and existing infrastructure, instead of building an entirely new end-to-end codegen from scratch.

Unless we can do something better by working natively with Google Discovery format, but I'm not really convinced of that.

**Comment by hamnis** [comment]
> but aren't you unnecessary replicating that toolchain?

Perhaps, but its was easier for me to do this.

Not sure if its worth the effort to look into guardrail, since I tried to do that before and I had to step away because of the sheer complexity.

**Comment by armanbilge** [comment]
> I had to step away because of the sheer complexity.

Yeah, that's fair enough. If Google is only using a subset of OpenAPI features then you may very well be right we don't need to deal with all that complexity.

**Comment by armanbilge** [comment]
Took a look at the generated code from your plugin. Very nice 👏

**Comment by hamnis** [comment]
I have finally been able to publish the first version.

`addSbtPlugin("net.hamnaberg" % "google-discovery-sbt" % "0.1.0")`

**Comment by armanbilge** [comment]
So actually what I'd like to do is use your plugin to add more modules to our googleapis-http4s repo.

https://github.com/davenverse/googleapis-http4s
